### PR TITLE
Wrapper Store

### PR DESCRIPTION
**Pull request by sstyer**

### CODE REVIEW

File: service/src/main/scala/org/oseraf/bullseye/service/DataService/DukeResolver.scala [hunk 1]
```diff
@@ -48,6 +62,7 @@ class DukeResolver(
   val dukeProcessor = new Processor(dukeConf)
 
   def deduplicate(): Seq[BullsEyeDedupeCandidate] = {
+    var cnt = -1
```
Comment (sumidiot) [inline service/src/main/scala/org/oseraf/bullseye/service/DataService/DukeResolver.scala]: Unused

File: service/src/main/scala/org/oseraf/bullseye/service/DataService/DukeResolver.scala [hunk 2]
```diff
@@ -73,7 +88,7 @@ class DukeResolver(
       None
     } else {
       val score = dukeProcessor.compare(targetRecord, candidateRecord)
-      logger.trace("Got comparison score " + score + " for " + candidateRecord.getValue("id"))
```
Comment (sumidiot) [inline service/src/main/scala/org/oseraf/bullseye/service/DataService/DukeResolver.scala]: Why hard-code "id" here, when ID_ATTRIBUTE is used elsewhere?

File: service/src/main/scala/org/oseraf/bullseye/service/DataService/DukeResolver.scala [hunk 3]
```diff
@@ -20,21 +21,34 @@ class DukeResolver(
 
   val keeperProps = dukeConf.getProperties.map(_.getName)
   logger.info("Indexing graph in Duke, properties: " + keeperProps.mkString(", "))
+  private var indexCount = 0
   val innerDB = {
     val db = new LuceneDatabase()
     db.setConfiguration(dukeConf)
-    store.entities.foreach(eid => db.index(new EntityRecord(eid, store.entity(eid))))
+    store.entities.foreach(eid => {
+      Try {
```
Comment (sumidiot) [inline service/src/main/scala/org/oseraf/bullseye/service/DataService/DukeResolver.scala]: These error-handling blocks should be moved out to nice tidy little pr

File: service/src/main/scala/org/oseraf/bullseye/service/DataService/GraphLoader.scala [hunk 4]
```diff
@@ -8,7 +8,7 @@ import org.oseraf.bullseye.service.Service
 
 import scala.collection.JavaConversions._
 
-trait GraphLoader extends Service {
+object GraphLoader extends Service {
```
Comment (sumidiot) [inline service/src/main/scala/org/oseraf/bullseye/service/DataService/GraphLoader.scala]: GraphLoader doesn't really extend Service, usefully, does it?

File: service/src/main/scala/org/oseraf/bullseye/service/Service.scala [hunk 5]
```diff
@@ -1,7 +1,20 @@
 package org.oseraf.bullseye.service
 
 import com.typesafe.config.ConfigFactory
+import java.io.File
+import com.typesafe.scalalogging.slf4j.Logging
 
-trait Service {
-  val conf = ConfigFactory.load().getConfig("bullseye").getConfig("service")
+trait Service extends Logging {
+  private val baseConfig = {
```
Comment (sumidiot) [inline service/src/main/scala/org/oseraf/bullseye/service/Service.scala]: I finally figured out how to do this more like "the right way". Service should be like it is in the master branch. Overrides can then be provided by, for example, setting -Dconfig.file=/path/to/whatever.conf, which can be achieved in tomcat (for example) with CATALINA_OPTS, among other methods.

File: store/src/main/scala/org/oseraf/bullseye/store/AttributeBasedSearchPlugin.scala [hunk 6]
```diff
@@ -52,13 +53,22 @@ trait BruteForceAttributeBasedNaivelyFuzzySearchPlugin
     else entity.attribute(key, default)
 }
 
-trait IndexedBlueprintsFuzzyVertexSearchPlugin extends AttributeBasedNaivelyFuzzySearchPlugin with BruteForceAttributeBasedNaivelyFuzzySearchPlugin {
-  val store:BlueprintsGraphStore with EntityIterationPlugin
+trait IndexedBlueprintsFuzzyVertexSearchPlugin extends AttributeBasedNaivelyFuzzySearchPlugin with BruteForceAttributeBasedNaivelyFuzzySearchPlugin with Logging  {
+  val store:BlueprintsGraphStore
+  def doBF: Boolean = true
```
Comment (sumidiot) [inline store/src/main/scala/org/oseraf/bullseye/store/AttributeBasedSearchPlugin.scala]: let's use a more descriptive variable name here, something like "bruteForceFallback" or something.

File: store/src/main/scala/org/oseraf/bullseye/store/AttributeBasedSearchPlugin.scala [hunk 7]
```diff
@@ -52,13 +53,22 @@ trait BruteForceAttributeBasedNaivelyFuzzySearchPlugin
...
   override def search(key: AttributeContainer.KEY, value:AttributeContainer.VALUE):Iterable[(EntityStore.ID, Double)] = {
-    store.graph.query().has(key, CONTAINS, value).vertices().toList.map(v => (v.getId.toString, score(v.getProperty(key), value)))
+    logger.trace("Trying indexed search")
+    val verts = store.graph.query().has(key, CONTAINS, value).vertices().toList
```
Comment (sumidiot) [inline store/src/main/scala/org/oseraf/bullseye/store/AttributeBasedSearchPlugin.scala]: This is somewhat unfortunate, in that CONTAINS is from Titan, and seems to have more like "EQUALS" meaning for non-Titan graphs. I don't see, however, how to do a CONTAINS search in a Tinkergraph, e.g.

File: store/src/main/scala/org/oseraf/bullseye/store/WriteEventPublisherPlugin.scala [hunk 8]
```diff
@@ -9,7 +9,7 @@ trait WriteEventPublisherPlugin extends EntityStore {
   var writeListeners = new mutable.ArrayBuffer[WriteEventListener]()
 
   abstract override def addEntity(id: EntityStore.ID, entity: Entity): Boolean =
-    returning(super.addEntity(id, entity)) { writeListeners.foreach(_.handleAddEntityEvent(id)) }
+    returning(super.addEntity(id, entity)) { writeListeners.foreach(_.handleAddEntityEvent(id, entity)) }
```
Comment (sumidiot) [inline store/src/main/scala/org/oseraf/bullseye/store/WriteEventPublisherPlugin.scala]: It seems like if we're changing the interface to pass the entity for this method, we should change all the other methods similarly. Maybe provide pairs of methods, so addEntity(id) and addEntity(id, entity), with a default implementation of addEntity(id, entity) = addEntity(id). This can/should be a separate pr

File: store/src/main/scala/org/oseraf/bullseye/store/impl/blueprints/BlueprintsGraphStore.scala [hunk 9]
```diff
@@ -1,5 +1,6 @@
 package org.oseraf.bullseye.store.impl.blueprints
 
+import com.thinkaurelius.titan.core.TitanGraph
```
Comment (sumidiot) [inline store/src/main/scala/org/oseraf/bullseye/store/impl/blueprints/BlueprintsGraphStore.scala]: Unused import

File: store/src/main/scala/org/oseraf/bullseye/store/impl/blueprints/BlueprintsGraphStore.scala [hunk 10]
```diff
@@ -57,6 +58,13 @@ class BlueprintsGraphStore(val graph: Graph)
     true
   }
 
+  def addEntityIfNecessary(id:EntityStore.ID, entity:Entity):Boolean = {
+    if(!store.entityExists(id)) {
+      store.addEntity(id, entity)
+      return true
+    }
+    return false
```
Comment (sumidiot) [inline store/src/main/scala/org/oseraf/bullseye/store/impl/blueprints/BlueprintsGraphStore.scala]: You don't need explicit returns in this method

File: store/src/main/scala/org/oseraf/bullseye/store/impl/blueprints/BlueprintsGraphStore.scala [hunk 11]
```diff
@@ -57,6 +58,13 @@ class BlueprintsGraphStore(val graph: Graph)
     true
   }
 
+  def addEntityIfNecessary(id:EntityStore.ID, entity:Entity):Boolean = {
+    if(!store.entityExists(id)) {
+      store.addEntity(id, entity)
+      return true
```
Comment (sumidiot) [inline store/src/main/scala/org/oseraf/bullseye/store/impl/blueprints/BlueprintsGraphStore.scala]: You should just return the return value of store.addEntity, which means you can remove this line.

File: store/src/main/scala/org/oseraf/bullseye/store/impl/blueprints/BlueprintsGraphStore.scala [hunk 12]
```diff
@@ -57,6 +58,13 @@ class BlueprintsGraphStore(val graph: Graph)
     true
   }
 
+  def addEntityIfNecessary(id:EntityStore.ID, entity:Entity):Boolean = {
+    if(!store.entityExists(id)) {
+      store.addEntity(id, entity)
```
Comment (sumidiot) [inline store/src/main/scala/org/oseraf/bullseye/store/impl/blueprints/BlueprintsGraphStore.scala]: Why "store." in these two lines? Isn't store=this?

File: webapp/src/main/scala/ScalatraBootstrap.scala [hunk 13]
```diff
@@ -16,7 +16,7 @@ class DefaultServlet extends ScalatraServlet with ScalateSupport {
 class ScalatraBootstrap extends LifeCycle {
   override def init(context: ServletContext) {
     context.mount(new DefaultServlet, "/", "bullseye")
-    context.mount(new DataAPI {}, "/rest/data", "theseus/data")
+    context.mount(new DataAPI {}, "/rest/data", "bullseye/data")
```
Comment (sumidiot) [inline webapp/src/main/scala/ScalatraBootstrap.scala]: We can/should make a little merge into master for this change.

File: webapp/src/main/ui/src/app/home/home.js [hunk 14]
```diff
@@ -75,7 +75,7 @@ angular.module('bullseye.home', [])
                 var nodes = [];
                 $scope.selected.items.forEach(function(n) {
                     n.entities.forEach(function (e){
-                        nodes.push({entity: e, score: Math.round(100.0 * n.score)});
+                        nodes.push({entity: e, score: Math.round(n.score)});
```
Comment (sumidiot) [inline webapp/src/main/ui/src/app/home/home.js]: We can/should make a little merge into master for this change.

File: ikanow/pom.xml [hunk 15]
```diff
@@ -56,6 +56,11 @@
 
     <dependencies>
         <dependency>
+            <groupId>commons-codec</groupId>
```
Comment (sumidiot) [inline ikanow/pom.xml]: What's this dependency for? Things seem to compile without it.

**Comment by sstyer** [comment]
So this PR (and the original intention) is very specific to a particular data-source, more than structurally to the schema -- the keys and the values specifically. You can see in the diff that I changed the keys and values to lowercase to fit another schema. I think the remaining work to be done related to this PR is abstracting the specific keys out.

**Comment by nhambletCCRI** [comment]
Replaced by #40
